command/agent/template/template_test.go

```diff
@@ -0,0 +1,156 @@
+package template
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	ctconfig "github.com/hashicorp/consul-template/config"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/command/agent/config"
+	"github.com/hashicorp/vault/helper/testhelpers"
+	"github.com/hashicorp/vault/sdk/helper/logging"
+)
+
+// TestNewServer is a simple test to make sure NewServer returns a Server and
+// channel
+func TestNewServer(t *testing.T) {
+	ts, ch := NewServer(&ServerConfig{})
+	if ts == nil {
+		t.Fatal("nil server returned")
+	}
+	if ch == nil {
+		t.Fatal("nil blocking channel returned")
+	}
+}
+
+func TestServerRun(t *testing.T) {
+	// create http test server
+	ts := httptest.NewServer(http.HandlerFunc(handleRequest))
+	defer ts.Close()
+	tmpDir, err := ioutil.TempDir("", "agent-tests")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testCases := map[string]struct {
+		templates []*ctconfig.TemplateConfig
+	}{
+		"basic": {
+			templates: []*ctconfig.TemplateConfig{
+				&ctconfig.TemplateConfig{
+					Contents: testhelpers.StrPtr(templateContents),
+				},
+			},
+		},
+	}
+
+	templateTokenCh := make(chan string, 1)
+	var unblockCh <-chan struct{}
+
+	// secretRender is a simple struct that represents the secret we render to
+	// disk. It's used to unmarshal the file contents and test against
+	type secretRender struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+		Version  string `json:"version"`
+	}
+
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			for i, template := range tc.templates {
+				// template.Destination = testhelpers.StrPtr(fmt.Sprintf("%s/render_%d.txt", tmpDir, i))
+				dstFile := fmt.Sprintf("%s/render_%d.txt", tmpDir, i)
+				template.Destination = testhelpers.StrPtr(dstFile)
+
+				ctx, cancelFunc := context.WithCancel(context.Background())
+				sc := ServerConfig{
+					Logger: logging.NewVaultLogger(hclog.Trace),
+					VaultConf: &config.Vault{
+						Address: ts.URL,
+					},
+				}
+
+				var server *Server
+				server, unblockCh = NewServer(&sc)
+				if ts == nil {
+					t.Fatal("nil server returned")
+				}
+				if unblockCh == nil {
+					t.Fatal("nil blocking channel returned")
+				}
+
+				go server.Run(ctx, templateTokenCh, tc.templates)
+
+				// send a dummy value to trigger the internal Runner to query for secret
+				// info
+				templateTokenCh <- "test"
+
+				// Unblock should close immediately b/c there are no templates to render
+				select {
+				case <-ctx.Done():
+				case <-unblockCh:
+				}
+
+				// cancel to clean things up
+				cancelFunc()
```

Comment:
I didnt defer because I wanted the temporary directory cleaned up. Defer would just cause all that to pile up until the entire test finished, correct?

Reply:
Aside, this made me realize I had structured this test poorly; I was testing the rendering of all templates in a loop of each template. If the test had multiple templates this probably would have been noticeable, but it only has 1 at the moment. I updated that in 05a85fea6 , and it will need further updates (the value checking is hard coded to this single template), but that can come in a follow-up PR if you're not opposed to it